File: lib/prorate/leaky_bucket.rb
module Prorate

  # This offers just the leaky bucket implementation with fill control, but without the timed lock.
  # It does not raise any exceptions, it just tracks the state of a leaky bucket in Redis.
  #
  # Important differences from the more full-featured Throttle class are:
  #
  # * No logging (as most meaningful code lives in Lua anyway)
  # * No timed block - if you need to keep track of timed blocking it can be done externally
  # * Leak rate is specified directly in tokens per second, instead of specifying the block period.
  # * The bucket level is stored and returned as a Float which allows for finer-grained measurement,
  #   but more importantly - makes testing from the outside easier.
  #
  # It does have a few downsides compared to the Throttle though
  #
  # * Bucket is only full momentarily. On subsequent calls some tokens will leak already, so you either
  #   need to do delta checks on the value or rely on putting the token into the bucket.
  class LeakyBucket
    LUA_SCRIPT_CODE = File.read(File.join(__dir__, "leaky_bucket.lua"))
    LUA_SCRIPT_HASH = Digest::SHA1.hexdigest(LUA_SCRIPT_CODE)

    class BucketState < Struct.new(:level, :full)
      # Returns the level of the bucket after the operation on the LeakyBucket
      # object has taken place. There is a guarantee that no tokens have leaked
      # from the bucket between the operation and the freezing of the BucketState
      # struct.
      #
      # @!attribute [r] level
      #   @return [Float]

      # Tells whether the bucket was detected to be full when the operation on
      # the LeakyBucket was performed. There is a guarantee that no tokens have leaked
      # from the bucket between the operation and the freezing of the BucketState
      # struct.
      #
      # @!attribute [r] full
      #   @return [Boolean]

      alias_method :full?, :full

      # Returns the bucket level of the bucket state as a Float
      #
      # @return [Float]
      def to_f
        level.to_f
      end

      # Returns the bucket level of the bucket state rounded to an Integer
      #
      # @return [Integer]
      def to_i
        level.to_i
      end
    end

    # Creates a new LeakyBucket. The object controls 2 keys in Redis: one
    # for the last access time, and one for the contents of the key.
    #
    # @param redis_key_prefix[String] the prefix that is going to be used for keys.
    #   If your bucket is specific to a user, a browser or an IP address you need to mix in
    #   those values into the key prefix as appropriate.
    # @param leak_rate[Float] the leak rate of the bucket, in tokens per second
    # @param redis[Redis,#with] a Redis connection or a ConnectionPool instance
    #   if you are using the connection_pool gem. With a connection pool Prorate will
    #   checkout a connection using `#with` and check it in when it's done.
    # @param bucket_capacity[Numeric] how many tokens is the bucket capped at.
    #   Filling up the bucket using `fillup()` will add to that number, but
    #   the bucket contents will then be capped at this value. So with
    #   bucket_capacity set to 12 and a `fillup(14)` the bucket will reach the level
    #   of 12, and will then immediately start leaking again.
    def initialize(redis_key_prefix:, leak_rate:, redis:, bucket_capacity:)
      @redis_key_prefix = redis_key_prefix
      @redis = NullPool.new(redis) unless redis.respond_to?(:with)
      @leak_rate = leak_rate.to_f
      @capacity = bucket_capacity.to_f
    end

    # Places `n` tokens in the bucket.
    #
    # @return [BucketState] the state of the bucket after the operation
    def fillup(n_tokens)
      run_lua_bucket_script(n_tokens.to_f)
    end

    # Returns the current state of the bucket, containing the level and whether the bucket is full
    #
    # @return [BucketState] the state of the bucket after the operation
    def state
      run_lua_bucket_script(0)
    end

    # Returns the Redis key for the leaky bucket itself
    # Note that the key is not guaranteed to contain a value if the bucket has not been filled
    # up recently.
    #
    # @return [String]
    def leaky_bucket_key
      "#{@redis_key_prefix}.leaky_bucket.bucket_level"
    end

    # Returns the Redis key under which the last updated time of the bucket gets stored.
    # Note that the key is not guaranteed to contain a value if the bucket has not been filled
    # up recently.
    #
    # @return [String]
    def last_updated_key
      "#{@redis_key_prefix}.leaky_bucket.last_updated"
    end

    private

    def run_lua_bucket_script(n_tokens)
      @redis.with do |r|
        begin
          # The script returns a tuple of "whole tokens, microtokens"
          # to be able to smuggle the float across (similar to Redis TIME command)
          level_str, is_full_int = r.evalsha(
            LUA_SCRIPT_HASH,
            keys: [leaky_bucket_key, last_updated_key], argv: [@leak_rate, n_tokens, @capacity])
          BucketState.new(level_str.to_f, is_full_int == 1)
        rescue Redis::CommandError => e
          if e.message.include? "NOSCRIPT"
            # The Redis server has never seen this script before. Needs to run only once in the entire lifetime
            # of the Redis server, until the script changes - in which case it will be loaded under a different SHA
            r.script(:load, LUA_SCRIPT_CODE)
            retry
          else
            raise e
          end
        end
      end
    end
  end
end

File: lib/prorate/leaky_bucket.lua
-- Single threaded Leaky Bucket implementation (without blocking).
-- args: key_base, leak_rate, bucket_ttl, fillup. To just verify the state of the bucket leak_rate of 0 may be passed.
-- returns: the leve of the bucket in number of tokens

-- this is required to be able to use TIME and writes; basically it lifts the script into IO
redis.replicate_commands()

-- Keys are passed separately as Redis ensures that the keys live on the
-- same shard (or at least verifies for it). You have to pass all the keys you intend to touch.
local bucket_level_key = KEYS[1]
local last_updated_key = KEYS[2]

local leak_rate = tonumber(ARGV[1])
local fillup = tonumber(ARGV[2]) -- How many tokens this call adds to the bucket.
local bucket_capacity = tonumber(ARGV[3]) -- How many tokens is the bucket allowed to contain

-- Compute the key TTL for the bucket. We are interested in how long it takes the bucket
-- to leak all the way to 0, as this is the time when the values stay relevant. We pad with 1 second
-- to have a little cushion.
local key_lifetime = math.ceil((bucket_capacity / leak_rate) + 1)

-- Take a timestamp
local redis_time = redis.call("TIME") -- Array of [seconds, microseconds]
local now = tonumber(redis_time[1]) + (tonumber(redis_time[2]) / 1000000)

-- get current bucket level. The throttle key might not exist yet in which
-- case we default to 0
local bucket_level = tonumber(redis.call("GET", bucket_level_key)) or 0

-- ...and then perform the leaky bucket fillup/leak. We need to do this also when the bucket has
-- just been created because the initial fillup to add might be so high that it will
-- immediately overflow the bucket and trigger the throttle, on the first call.
local last_updated = tonumber(redis.call("GET", last_updated_key)) or now -- use sensible default of 'now' if the key does not exist

-- Subtract the number of tokens leaked since last call
local dt = now - last_updated
local new_bucket_level = bucket_level - (leak_rate * dt) + fillup

-- and _then_ and add the tokens we fillup with. Cap the value to be 0 < capacity
new_bucket_level = math.max(0, math.min(bucket_capacity, new_bucket_level))

-- Since we return a floating point number string-formatted even if the bucket is full we
-- have some loss of precision in the formatting, even if the bucket was actually full.
-- This bit of information is useful to preserve.
local at_capacity = 0
if new_bucket_level == bucket_capacity then
  at_capacity = 1
end

-- If both the initial level was 0, and the level after putting tokens in is 0 we
-- can avoid setting keys in Redis at all as this was only a level check.
if new_bucket_level == 0 and bucket_level == 0 then
  return {"0.0", at_capacity}
end

-- Save the new bucket level
redis.call("SETEX", bucket_level_key, key_lifetime, new_bucket_level)

-- Record when we updated the bucket so that the amount of tokens leaked
-- can be correctly determined on the next invocation
redis.call("SETEX", last_updated_key, key_lifetime, now)

-- Most Redis adapters when used with the Lua interface truncate floats
-- to integers (at least in Python that is documented to be the case in
-- the Redis ebook here
-- https://redislabs.com/ebook/part-3-next-steps/chapter-11-scripting-redis-with-lua/11-1-adding-functionality-without-writing-c
-- We need access to the bucket level as a float value since our leak rate might as well be floating point, and to achieve that
-- we can go two ways. We can turn the float into a Lua string, and then parse it on the other side, or we can convert it to
-- a tuple of two integer values - one for the integer component and one for fraction.
-- Now, the unpleasant aspect is that when we do this we will lose precision - the number is not going to be
-- exactly equal to capacity, thus we lose the bit of information which tells us whether we filled up the bucket or not.
-- Also since the only moment we can register whether the bucket is above capacity is now - in this script, since
-- by the next call some tokens will have leaked.
return {string.format("%.9f", new_bucket_level), at_capacity}
